vowpalwabbit/namespaced_features.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "namespaced_features.h"
+#include "vw_exception.h"
+
+using namespace VW;
+
+features* namespaced_features::get_feature_group(uint64_t hash)
+{
+  auto it = _hash_to_index_mapping.find(hash);
+  if (it == _hash_to_index_mapping.end()) { return nullptr; }
+
+  return &_feature_groups[it->second];
+}
+
+const features* namespaced_features::get_feature_group(uint64_t hash) const
+{
+  auto it = _hash_to_index_mapping.find(hash);
+  if (it == _hash_to_index_mapping.end()) { return nullptr; }
+
+  return &_feature_groups[it->second];
+}
+
+std::vector<namespace_index> namespaced_features::get_indices() const
+{
+  auto indices_copy = _namespace_indices;
+  std::sort(indices_copy.begin(), indices_copy.end());
+  auto last = std::unique(indices_copy.begin(), indices_copy.end());
+  indices_copy.erase(last, indices_copy.end());
+  return indices_copy;
+}
+
+std::pair<namespaced_features::indexed_iterator, namespaced_features::indexed_iterator>
+namespaced_features::get_namespace_index_groups(namespace_index ns_index)
+{
+  return std::make_pair(namespace_index_begin(ns_index), namespace_index_end(ns_index));
+}
+
+std::pair<namespaced_features::const_indexed_iterator, namespaced_features::const_indexed_iterator>
+namespaced_features::get_namespace_index_groups(namespace_index ns_index) const
+{
+  return std::make_pair(namespace_index_begin(ns_index), namespace_index_end(ns_index));
+}
+
+features& namespaced_features::get_or_create_feature_group(uint64_t hash, namespace_index ns_index)
+{
+  auto* existing_group = get_feature_group(hash);
+  if (existing_group == nullptr)
+  {
+    _feature_groups.emplace_back();
+    _namespace_indices.push_back(ns_index);
+    _namespace_hashes.push_back(hash);
+    auto new_index = _feature_groups.size() - 1;
+    _hash_to_index_mapping[hash] = new_index;
+    _legacy_indices_to_index_mapping[ns_index].push_back(new_index);
+    existing_group = &_feature_groups.back();
+  }
+
+  return *existing_group;
+}
+
+const features& namespaced_features::operator[](uint64_t hash) const
```

Comment:
Many believe that map's get or create with `[]` is not exactly good API design though. That being said though I've been struggling to find the right ergonomics for these `get` and `create` functions

Reply:
The tricky part is that to create both hash and index are required to create a feature_group (since the merged hash and index proposal was put off due to it being a breaking change)